state-reconstruct-fetcher/src/l1_fetcher.rs

```diff
@@ -222,14 +232,42 @@ impl L1Fetcher {
                 let mut latest_l2_block_number = U256::zero();
                 let mut previous_hash = None;
                 loop {
-                    // Break when reaching the `end_block` or on the receivement of a `ctrl_c` signal.
-                    if (disable_polling && current_l1_block_number > end_block_number)
-                        || cancellation_token.is_cancelled()
-                    {
-                        tracing::debug!("Shutting down main handle...");
+                    // Break on the receivement of a `ctrl_c` signal.
+                    if cancellation_token.is_cancelled() {
+                        tracing::debug!("Shutting down main handler...");
                         break;
                     }
 
+                    let Some(end_block_number) = end_block else {
+                        if let Ok(new_end) = L1Fetcher::retry_call(
+                            || provider_clone.get_block(BlockNumber::Latest),
+                            L1FetchError::GetEndBlockNumber,
+                        )
+                        .await
+                        {
+                            if let Some(found_block) = new_end {
+                                if let Some(end_block_number) = found_block.number {
+                                    end_block = Some(end_block_number);
+                                    metrics.lock().await.last_l1_block = end_block_number.as_u64();
+                                }
+                            }
+                        } else {
+                            tracing::debug!("Cannot get latest block number...");
+                            tokio::time::sleep(Duration::from_secs(LONG_POLLING_INTERVAL_S)).await;
+                        }
+
+                        continue;
+                    };
+
+                    if current_l1_block_number > end_block_number {
+                        // This function must not be called w/ current_l1_block_number > end_block_number .
+                        assert!(!disable_polling);
```

Comment:
```suggestion
                    if current_l1_block_number > end_block_number && !disable_polling {
```
Instead of panicking, we can have it as an additional condition

Reply:
This really shouldn't happen - any place that increases `current_l1_block_number` beyond `end_block_number` (there's only one) must check the `disabled_polling` case first. Plus, requiring `!disable_polling` to skip resetting the end block is wrong - in the `disable_polling` case, the code would just ask for future blocks, never ending and never getting any data. The correct reaction is to return from the function, and that already happened, elsewhere...

Reply:
well yes, that's what the assert says...

Reply:
Gotcha! I would update the comment then to briefly explain why we do the `assert!`, i.e:
> any place that increases current_l1_block_number beyond end_block_number (there's only one) must check the disabled_polling case first

Reply:
OK